### PR TITLE
Add 'volatile' to cpuid inline asm

### PR DESCRIPTION
**Pull request by thomcc**
`cpuid` is a serializing instruction, and is the only serializing instruction available as an intrinsic AFAIK. Various points of intels documentation suggests to use it as a way to fence other instructions — The typical case is in conjunction with rdtsc/rdtscp for measurement (though on sse2 processors you can also use `mfence`/`lfence` for this purpose), although this is not the only use case.

Without the "volatile" clobber, this intrinsic can't easily be used this way, since you have to fake the use of the result or LLVM will entirely remove it. This fails to reflect the behavior documentated for instruction, so it seems like a clear-cut bug worth fixing to me.

See these for further info if desired:

- https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-system-programming-manual-325384.pdf (in particular sections on serializing instructions)
- https://www.intel.com/content/dam/www/public/us/en/documents/white-papers/ia-32-ia-64-benchmark-code-execution-paper.pdf (which is dated and fails to mention lfence/mfence, but still applies for code in a "no sse2" fallback case)

TLDR: CPUID has side effects in the x86 memory model (including out of order execution) and this patch makes sure the intrinsic works for that.

### CODE REVIEW

**Comment by Amanieu** [comment]
I would prefer if that `llvm_asm!` was rewritten to use `asm!` instead. `llvm_asm!` is going to be deprecated soon and will eventually be removed from the compiler.

**Comment by thomcc** [comment]
Sure, didn't want to do that off the bat since it might have hidden the
change maybe.

Not super sure about the clobbers/options (still new to `asm!`), but this
seems to work and does prevent the removal as I hoped.

File: crates/core_arch/src/x86/cpuid.rs [hunk 1]
```diff
@@ -55,26 +55,33 @@ pub unsafe fn __cpuid_count(leaf: u32, sub_leaf: u32) -> CpuidResult {
...
-                  : "={eax}"(eax), "={ebx}"(ebx), "={ecx}"(ecx), "={edx}"(edx)
-                  : "{eax}"(leaf), "{ecx}"(sub_leaf)
-                  : :);
+        asm!(
+            "cpuid",
+            inlateout("eax") leaf => eax,
+            lateout("ebx") ebx,
+            inlateout("ecx") sub_leaf => ecx,
+            lateout("edx") edx,
+            options(nostack, nomem, preserves_flags),
```
Comment (thomcc) [inline crates/core_arch/src/x86/cpuid.rs:67]: Is an instruction that (among other things) acts as a fence actually `nomem`? IDK. I added it under the logic of: It shouldn't have a \"memory\" clobber.

But I suspect that's not quite what it means.

**Comment by sunfishcode** [comment]
Compilers don't consider out of order execution to be an observable side effect, so volatile won't be enough to prevent compilers from moving most instructions around it.

File: crates/core_arch/src/x86/cpuid.rs (hunk 1 above)
Comment (Amanieu) [reply crates/core_arch/src/x86/cpuid.rs:67]: You need to remove `nomem` for the `asm!` to act as a compiler fence. This will prevent memory accesses from being reordered across the `asm!`.

**Comment by thomcc** [comment]
> Compilers don't consider out of order execution to be an observable side effect, so volatile won't be enough to prevent compilers from moving most instructions around it.

I'm aware. I don't think that should be handled by this intrinsic, as there are other ways of dealing with that already.

File: crates/core_arch/src/x86/cpuid.rs (hunk 1 above)
Comment (thomcc) [reply crates/core_arch/src/x86/cpuid.rs]: Yeah, thanks.
